doc/storagemond/storagemond-hld.md

```diff
@@ -0,0 +1,452 @@
+
+1. **Check for `psutil` Module**:
+   - The functions first check if the `psutil` module is available in the current environment by examining the `sys.modules` dictionary.
+
+2. **Use `psutil` Module (if available)**:
+   - If `psutil` is available:
+     - The functions retrieve disk I/O counters, specifying the disk for which to get the counters.
+     - They then get the read or write count for the specified disk using `read_count` or `write_count` respectively.
+
+3. **Fallback to Parsing Disk Stats File**:
+   - If `psutil` is not available:
+     - The functions open the `/proc/diskstats` file
+     - They read the contents of the file and iterate over each line.
+     - For each line, they check if the name of the storage disk is present.
+     - If the name of the storage disk is found in the line, they return the value at the appropriate zero-based index (3 for reads, 7 for writes).
+     - If no line contains the name of the storage disk, they save the respective values as 0.
+
+4. **Return to caller**:
+     - These values are then returned to the caller to subsequently be written to `STATE_DB`.
+     - These values are also used to determine the total number of procfs reads and writes as explained in the following section
+
+
+#### **2.4.4 Accounting for reboots and unintended powercycles**
+
+The reset of values in `/proc/diskstats` upon device reboot or power cycle presents a challenge for maintaining long-term data integrity. To mitigate this challenge, we propose the following design considerations:
+
+1. Introduction of a bind-mounted directory within the pmon container at `/usr/share/storagemon/` which maps to `/host/pmon/storagemon/` on the host:
+    - This directory hosts a file named `fsio-rw-stats.json`, where the following values are stored:
+        - **most recent** Total reads and writes from the `STATE_DB`.
+        - **most recent** procfs reads and writes from the `STATE_DB`.
+
+    - This file would be read by the daemon on initialization after a planned reboot of the system, or in a graceful `stormond` restart scenario.
+
+2. Implementation of a script, tentatively named `fsio-rw-sync`, to be invoked by SONiC's reboot utility:
+    - This script would be called by the several reboot scripts and would be responsible for parsing and storing the aforementioned paramenters just before OS-level reboot.
+    - This script provides a mechanism to sync the procFS reads and writes values from the `STATE_DB` before they are reset by the reboot.
+
+##### **2.4.4.1 Daemon Restart / Reboot / Unintended Powercycle Scenario Behaviors**
+
+1. **Planned cold, fast, and warm reboot scenario**
+
+    - Prior to invoking an OS-level reboot, the latest FSIO Read and Write metrics are captured from the `/proc/diskstats` file and stored into the `fsio-rw-stats.json` by executing the `fsio-rw-sync` script from the respective reboot script (cold, soft, or warm).
```

Comment:
Addressed in latest commit

Reply:
After discussion we have decided to remove the sync script and instead adopt the following strategy:

cold/soft-reboot: When reboot script sends SIGTERM to pmon which in turn sends SIGTERM to stormond, we catch that signal and sync the STATE_DB values to the JSON file.

warm-reboot: We add the `STORAGE_INFO|` key to the backup_database() function in the fast-reboot script so that the values survive the system reboot.

I will make above changes to HLD.